### PR TITLE
Submodule

### PR DESCRIPTION
**Pull request by noppanit**
Submodule workflow

### CODE REVIEW

File: wordpress-development/submodule-workflow.md [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+
+We use submodule to maintain our internal and external plugins. Here's the best [article](http://blogs.atlassian.com/2013/03/git-submodules-workflows-tips/) I have found so far. 
+
+## How to roll back submodule SHA1
+
+From time to time, especially if we have a lot of changes we might forget to not commit the SHA1. If you do already, here's the step to roll back before the PR can be merged. 
+
+1. You would see something like this. 
+
+![submodule conflict](https://dl.dropboxusercontent.com/u/27106646/Github-Images/Screenshot%202015-05-12%2018.40.22.png)
```
Comment (danielbachhuber) [inline wordpress-development/submodule-workflow.md]: Can you upload the image to GH?

**Comment by noppanit** [comment]
I have made mistakes about this all the time. So I think it's good that we have some reference so people can roll back just in case @danielbachhuber  @davisshaver  @goldenapples  Can you guys review.

File: wordpress-development/submodule-workflow.md (hunk 1 above)
Comment (noppanit) [inline wordpress-development/submodule-workflow.md]: That's what I'm wondering how do I do that? From my Google I have to create a new issue and upload the image to that issue? like this? http://solutionoptimist.com/2013/12/28/awesome-github-tricks/

Comment (danielbachhuber) [inline wordpress-development/submodule-workflow.md]: > From my Google I have to create a new issue and upload the image to that issue?

Yep, that's how I usually do it. Or some comment on a thread.

**Comment by goldenapples** [comment]
Definitely good to have a document on this!

Since its continually an issue, would it be useful to define an actual best practice as well?

I don't know what a best practice across-the-board would be. For myself I have the entire `lib` directory marked to never commit with `git update-index --assume-unchanged lib/*`. And I have a function defined in my dotfiles that checks out a new branch and updates submodules at the same time:

```
function go {
    git checkout $1 && git submodule update
}
```
so that `go master` checks out master branch and updates all submodules.

There are all kinds of dangers with this approach (having to remember if I have work in progress on any of the submodules before changing branches, for one), so I don't know that I would recommend it to everyone, but it would be good to have a best practice to recommend for new devs, as git submodules aren't something lots of people are used to dealing with. I liked the suggestion of adding `ignore = dirty` to the .gitmodules file that Atlassian blog post recommended.

**Comment by noppanit** [comment]
Thanks for the input @goldenapples I'll update the doc. I think it's a good idea to have different options so people can choose to do what they're comfortable with.

**Comment by noppanit** [comment]
@danielbachhuber @goldenapples #reviewmerge

**Comment by danielbachhuber** [comment]
@noppanit I made a few edits in 79295e9bfaee41796fc46ff26c045f92ae4bb50d

Given the difficulty @lisaej was having with submodules, I'll leave to her for the final #reviewmerge, as she can likely validate whether this is a good introductory guide :)

**Comment by noppanit** [comment]
Sounds good.

**Comment by danielbachhuber** [comment]
Bump @lisaej

File: wordpress-development/submodule-workflow.md [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+1. You would see something like this. 
+
+![screenshot 2015-05-12 18 40 22](https://cloud.githubusercontent.com/assets/638379/7600363/6e8d94bc-f8d8-11e4-910a-d03abe6b8a2c.png)
+
+which you don't want it to happen in your PR. 
+
+2. So to rollback the hash you need to go into the directory of that submodule and do `git checkout 1dba7d7ad049c06ffc7a35a6d65d4ff4098a5fcd` (according to the example above) which will checkout the previous version of the submodule. 
+
+3. Then you can commit the change. `git add -A; git commit -m "advance submodule to previous revision"`
```
Comment (lisaej) [inline wordpress-development/submodule-workflow.md]: Do you still do this from submodule directory? Or from fusion-theme?

File: wordpress-development/submodule-workflow.md [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+function go {
+    git checkout $1 && git submodule update
+}
+```
+
+So when you want to checkout a branch you can just do `go master`
+
+## GIT Gui
+
+If you're using Git GUI you can also do this. [Git submodule gui](http://stackoverflow.com/questions/12899163/where-is-git-submodule-update-in-sourcetree)
```
Comment (lisaej) [inline wordpress-development/submodule-workflow.md]: This one is for sourcetree, I think. It doesn't exist for Github Mac.

**Comment by danielbachhuber** [comment]
Guess this is done? :smile:

**Pull request merged**